### PR TITLE
Create SampleQuerySet class 

### PR DESCRIPTION
**Pull request by hannako**
First step in the implementation of an automated evaluation workflow.

Based on @csutter proof of concept branch: https://github.com/alphagov/search-api-v2/compare/evaluation

Pair programming effort with @csutter and @leenagupte 🥧 🟣 🪄 

[Trello](https://trello.com/c/FVVlDFbV/842-evaluations-create-monthly-sample-query-set-from-clickstream-data)

### CODE REVIEW

File: app/clients/discovery_engine/services.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require "google/cloud/discovery_engine/v1beta"
+
+module DiscoveryEngine
+  module Services
```
[the remark this answers is not in the document]
Comment (hannako) [reply app/clients/discovery_engine/services.rb:4]: Yep - had a similar thought - I'll make that change 👍

File: app/services/discovery_engine/evaluation/sample_query_set.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+      def create
+        import_data(empty_query_set.name)
+      end
+
+    private
+
+      attr_reader :year, :month, :sqss_client, :sqs_client
+
+      def empty_query_set
```
Comment (csutter) [inline app/services/discovery_engine/evaluation/sample_query_set.rb:24]: I reckon either the whole GCP query set or at least its name should form part of the object's state - that way we wouldn't have to pass the name separately into the `import_data` method.

Given that objects of this class represent a query set on GCP, would it perhaps make more sense to have a factory method that creates the query set on GCP, and then returns an instance associated with that remote resource?

So you'd go through something like this in the ultimate caller (Rake task):
```ruby
sqs = DiscoveryEngine::Evaluation::SampleQuerySet.create(2025, 5)
sqs.import_from_bigquery(dataset: "automated_evaluation_input", table: "clickstream")
```

File: app/services/discovery_engine/evaluation/sample_query_set.rb [hunk 3]
```diff
@@ -26,6 +26,30 @@ def create_empty_set
           )
       end
 
+      def import_from_bigquery(dataset_id:, table_id:)
```
Comment (leenagupte) [inline app/services/discovery_engine/evaluation/sample_query_set.rb:29]: Minor: Perhaps `dataset_id` and `table_id` have default values? Unless we expect the names to be configurable in the rake task?

File: spec/services/discovery_engine/evaluation/sample_query_set_spec.rb [hunk 4]
```diff
@@ -3,10 +3,13 @@
 
   let(:month) { "1" }
   let(:year) { "2025" }
-  let(:sample_query_set_service_stub) { double("sample_query_set_service", create_sample_query_set: true) }
+  let(:operation_object) { double("operation", wait_until_done!: true, error?: false) }
```
Comment (leenagupte) [inline spec/services/discovery_engine/evaluation/sample_query_set_spec.rb:6]: Oof! I hadn't realised that you'd need to do this. Nice!

File: spec/lib/tasks/evaluation_spec.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+RSpec.describe "Evaluation tasks" do
+  describe "evaluation:clickstream:setup_sample_set" do
+    let(:sample_set) { class_double(DiscoveryEngine::Evaluation::SampleQuerySet) }
+    let(:sample_query_set_service_stub) { double("sample_query_set_service", create_sample_query_set: double(name: "name")) }
+    let(:sample_query_service_stub) { double("sample_query_service", import_sample_queries: double(wait_until_done!: true, error?: false)) }
+
+    before do
+      allow(DiscoveryEngine::Clients)
```
Comment (csutter) [inline spec/lib/tasks/evaluation_spec.rb:8]: I reckon this is an implementation detail of `DiscoveryEngine::Evaluation::SampleQuerySet`, we probably shouldn't need to stub that here.

File: spec/lib/tasks/evaluation_spec.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+          sample_query_set_service: sample_query_set_service_stub,
+          sample_query_service: sample_query_service_stub,
+        )
+      Rake::Task["evaluation:clickstream:setup_sample_set"].reenable
+    end
+
+    it "creates and imports a sample set" do
+      Timecop.freeze(Time.zone.local(2024, 12, 13, 1, 2, 3)) do
+        expect(sample_set)
+          .to receive(:create)
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply spec/lib/tasks/evaluation_spec.rb:12]: Thanks! This worked. I've updated and fixed up.

File: app/services/discovery_engine/evaluation/sample_query_set.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
+module DiscoveryEngine
+  module Evaluation
+    class SampleQuerySet
+      def initialize(month, year)
+        @month = month
+        @year = year
+        @set = create_empty_set
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/discovery_engine/evaluation/sample_query_set.rb]: I've added a commit to try this out and see what it looks like.

File: app/services/discovery_engine/evaluation/sample_query_set.rb (hunk 3 above)
Comment (leenagupte) [reply app/services/discovery_engine/evaluation/sample_query_set.rb]: I've added a couple of commits to try this.

File: app/services/discovery_engine/evaluation/sample_query_set.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      def initialize(month, year, set, table_id)
+        @month = month
+        @year = year
+        @set = set
+        @table_id = table_id
+      end
+
+      attr_reader :set, :table_id
+
+      def self.create(table_id: BIGQUERY_TABLE_ID)
```
[the remark this answers is not in the document]
Comment (hannako) [reply app/services/discovery_engine/evaluation/sample_query_set.rb:16]: I thought we decided that we would hardcode dates, but for table_id we'd have a default and allow it to be customised.

File: spec/services/discovery_engine/evaluation/sample_query_set_spec.rb [hunk 9]
```diff
@@ -34,15 +39,15 @@
         table_id: described_class::BIGQUERY_TABLE_ID,
         project_id: Rails.application.config.google_cloud_project_id,
         partition_date: {
-          year: year,
-          month: month,
+          year: current_year,
```
Comment (leenagupte) [inline spec/services/discovery_engine/evaluation/sample_query_set_spec.rb:62]: We should also test the January edge case where the previous month is also the previous year.

File: app/services/discovery_engine/evaluation/sample_query_set.rb [hunk 10]
```diff
@@ -12,7 +12,9 @@ def initialize(month, year, set)
 
       attr_reader :set
 
-      def self.create(month:, year:)
+      def self.create
+        month = Time.zone.now.prev_month.month
+        year = Time.zone.now.prev_month.year
```
Comment (leenagupte) [inline app/services/discovery_engine/evaluation/sample_query_set.rb:18]: Oooh. I didn't know there was a `prev_month` helper method. I've always used `1.months.ago`.

File: lib/tasks/evaluation.rake [hunk 11]
```diff
@@ -0,0 +1,11 @@
+namespace :evaluation do
+  namespace :clickstream do
+    desc "Create a sample query set for last month's clickstream data and import from BigQuery"
+    task :setup_sample_set, [:table_id] => :environment do |_, args|
+      args.with_defaults(table_id: DiscoveryEngine::Evaluation::SampleQuerySet::BIGQUERY_TABLE_ID)
```
Comment (leenagupte) [inline lib/tasks/evaluation.rake:5]: This is nice! I didn't know about `with_defaults`

**Review by leenagupte — APPROVED** [review]
It looks great! Thanks for adding links in the commit messages to explain what everything is and why it's needed. ⭐
